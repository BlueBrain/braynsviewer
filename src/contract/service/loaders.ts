export type LoaderPropertyTypeDefinition =
    | "string"
    | "number"
    | "boolean"
    | "integer"
    // An array of exactly 3 floats.
    | "vector3"
    // An array of exactly 4 floats.
    | "vector4"
    // An enumerate of strings.
    | string[]
    | "integer[]" // An array of integers. Look at `minItems` and `maxItems` for size.
    | "number[]" // An array of floats. Look at `minItems` and `maxItems` for size.
    | "string[]" // An array of strings. Look at `minItems` and `maxItems` for size.

export interface LoaderPropertyDefinition {
    // Atribute name (ex: 002RandomSeed).
    name: string
    // Display name (ex: Random seed for target subsetting).
    title: string
    defaultValue?: unknown
    description?: string
    type?: LoaderPropertyTypeDefinition
    minItems?: number
    maxItems?: number
}

export interface LoaderDefinition {
    name: string
    extensions: string[]
    properties: LoaderPropertyDefinition[]
}

export interface AddModelParams {
    bounding_box?: boolean
    loader_name: string
    loader_properties?: { [key: string]: unknown }
    name?: string
    path: string
    transformation?: {
        rotation?: [number, number, number, number]
        rotation_center?: [number, number, number]
        scale?: [number, number, number]
        translation?: [number, number, number]
    }
    visible?: boolean
}

/**
 * Get information on available loaders.
 */
export default interface LoadersServiceInterface {
    listAvailableLoaders(): Promise<LoaderDefinition[]>
    /**
     *
     * @param params
     */
    addModel(params: AddModelParams): Promise<unknown>
}
