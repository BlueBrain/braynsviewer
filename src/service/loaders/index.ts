export { default } from "./loaders-service"
