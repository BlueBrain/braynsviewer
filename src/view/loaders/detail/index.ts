export { default } from "./loader-detail-view"
