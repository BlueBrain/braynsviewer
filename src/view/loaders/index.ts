export { default } from "./loaders-view"
