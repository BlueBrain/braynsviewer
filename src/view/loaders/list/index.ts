export { default } from "./loaders-list-view"
