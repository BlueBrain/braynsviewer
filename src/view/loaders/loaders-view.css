.view-LoadersView > div.children {
    padding: 0.5rem;
}

.view-LoadersView div.error {
    padding: 1rem;
    border-radius: 0.5rem;
    background-color: var(--theme-color-error);
    color: var(--theme-color-on-error);
    margin: 1rem;
}
