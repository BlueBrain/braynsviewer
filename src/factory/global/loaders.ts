import { makeBraynsService } from "."
import { BraynsServiceAddress } from "../../contract/service/brayns"
import LoadersServiceInterface from "../../contract/service/loaders"
import LoadersService from "../../service/loaders"

export default function makeLoadersService(
    braynsAddress: BraynsServiceAddress
): LoadersServiceInterface {
    return new LoadersService(makeBraynsService(braynsAddress))
}
