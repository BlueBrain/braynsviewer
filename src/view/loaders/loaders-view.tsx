import LoadersServiceInterface, {
    LoaderDefinition,
} from "@/contract/service/loaders"
import Wizard from "@/ui/view/wizard"
import * as React from "react"
import LoaderDetail from "./detail"
import LoadersList from "./list"
import "./loaders-view.css"

export interface LoadersViewProps {
    className?: string
    loadersService: LoadersServiceInterface
}

const STEP_LIST = "list"
const STEP_DETAIL = "detail"

export default function LoadersView(props: LoadersViewProps) {
    const { loadersService } = props
    const [selectedLoader, setSelectedLoader] = React.useState<
        LoaderDefinition | undefined
    >(undefined)
    const [step, setStep] = React.useState(STEP_LIST)
    return (
        <Wizard step={step} className={getClassNames(props)}>
            <LoadersList
                key={STEP_LIST}
                loadersService={loadersService}
                onSelect={(loader) => {
                    setSelectedLoader(loader)
                    setStep(STEP_DETAIL)
                }}
            />
            <LoaderDetail
                key={STEP_DETAIL}
                loader={selectedLoader}
                loadersService={loadersService}
                onBack={() => {
                    setStep(STEP_LIST)
                }}
            />
        </Wizard>
    )
}

function getClassNames(props: LoadersViewProps): string {
    const classNames = ["custom", "view-LoadersView"]
    if (typeof props.className === "string") {
        classNames.push(props.className)
    }

    return classNames.join(" ")
}
