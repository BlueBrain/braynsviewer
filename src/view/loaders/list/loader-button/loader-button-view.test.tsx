// To test a React component, you need to install these modules:
// yarn add --dev react-test-renderer @types/react-test-renderer
// @see https://jestjs.io/docs/en/snapshot-testing
//
// If a test failed just because you intended to improve the component,
// just call `jest --updateSnapshot`.

import { LoaderDefinition } from "@/contract/service/loaders"
import React from "react"
import Renderer from "react-test-renderer"
import LoaderButtonView, { LoaderButtonViewProps } from "./loader-button-view"

function view(partialProps: Partial<LoaderButtonViewProps>) {
    const props: LoaderButtonViewProps = {
        value: {
            name: "mesh",
            extensions: [".obj", ".blend"],
            properties: [],
        },
        onClick() {},
        ...partialProps,
    }
    return Renderer.create(<LoaderButtonView {...props} />).toJSON()
}

describe("<LoaderButtonView/> in view/loaders", () => {
    it("should be consistent with previous snapshot", () => {
        expect(view({})).toMatchSnapshot()
    })
})
