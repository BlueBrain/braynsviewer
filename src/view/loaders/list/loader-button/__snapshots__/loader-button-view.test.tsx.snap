// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoaderButtonView/> in view/loaders should be consistent with previous snapshot 1`] = `
<button
  className="custom ui-view-TouchableView custom view-loaders-LoaderButtonView"
  onClick={[Function]}
>
  <div>
    mesh
  </div>
  <div
    title=".obj, .blend"
  >
    .obj, .blend
  </div>
</button>
`;
